### PR TITLE
Update RustPython to fix `Dict.keys` type

### PR DESCRIPTION
**Pull request by harupy**
This PR upgrades RustPython to fix the type of `Dict.keys` to `Vec<Option<Expr>>` (see https://github.com/RustPython/RustPython/pull/4449 for why this change was needed) and unblock #1884.

### CODE REVIEW

File: src/rules/pyupgrade/rules/convert_typed_dict_functional_to_class.rs [hunk 1]
```diff
@@ -78,14 +78,21 @@ fn create_class_def_stmt(
     })
 }
 
-fn properties_from_dict_literal(keys: &[Expr], values: &[Expr]) -> Result<Vec<Stmt>> {
+// let Some(key) = key else {
+//     bail!("Unpacking is not supported")
+// };
```
Comment (harupy) [inline src/rules/pyupgrade/rules/convert_typed_dict_functional_to_class.rs:83]:
```suggestion
```

File: src/rules/pyflakes/rules/repeated_keys.rs [hunk 2]
```diff
@@ -26,16 +26,16 @@ fn into_dictionary_key(expr: &Expr) -> Option<DictionaryKey> {
...
     // Generate a map from key to (index, value).
     let mut seen: FxHashMap<DictionaryKey, FxHashSet<ComparableExpr>> =
         FxHashMap::with_capacity_and_hasher(keys.len(), BuildHasherDefault::default());
 
     // Detect duplicate keys.
-    for (i, key) in keys.iter().enumerate() {
-        if let Some(key) = into_dictionary_key(key) {
-            if let Some(seen_values) = seen.get_mut(&key) {
-                match key {
+    for (i, key) in keys.iter().flatten().enumerate() {
```
Comment (harupy) [inline src/rules/pyflakes/rules/repeated_keys.rs:35]: Ah, we can't use flatten here.

**Comment by harupy** [comment]
Prior to https://github.com/RustPython/RustPython/pull/4449,

```python
{"a": 0, **b, "c": 1}
```

is parsed as:

```
[
    Located {
        location: Location {
            row: 1,
            column: 0,
        },
        end_location: Some(
            Location {
                row: 1,
                column: 21,
            },
        ),
        custom: (),
        node: Expr {
            value: Located {
                location: Location {
                    row: 1,
                    column: 0,
                },
                end_location: Some(
                    Location {
                        row: 1,
                        column: 21,
                    },
                ),
                custom: (),
                node: Dict {
                    keys: [
                        Located {
                            location: Location {
                                row: 1,
                                column: 1,
                            },
                            end_location: Some(
                                Location {
                                    row: 1,
                                    column: 4,
                                },
                            ),
                            custom: (),
                            node: Constant {
                                value: Str(
                                    "a",
                                ),
                                kind: None,
                            },
                        },
                    ],
                    values: [
                        Located {
                            location: Location {
                                row: 1,
                                column: 6,
                            },
                            end_location: Some(
                                Location {
                                    row: 1,
                                    column: 7,
                                },
                            ),
                            custom: (),
                            node: Constant {
                                value: Int(
                                    0,
                                ),
                                kind: None,
                            },
                        },
                        Located {
                            location: Location {
                                row: 1,
                                column: 11,
                            },
                            end_location: Some(
                                Location {
                                    row: 1,
                                    column: 12,
                                },
                            ),
                            custom: (),
                            node: Name {
                                id: "b",
                                ctx: Load,
                            },
                        },
                        Located {
                            location: Location {
                                row: 1,
                                column: 14,
                            },
                            end_location: Some(
                                Location {
                                    row: 1,
                                    column: 17,
                                },
                            ),
                            custom: (),
                            node: Dict {
                                keys: [
                                    Located {
                                        location: Location {
                                            row: 1,
                                            column: 14,
                                        },
                                        end_location: Some(
                                            Location {
                                                row: 1,
                                                column: 17,
                                            },
                                        ),
                                        custom: (),
                                        node: Constant {
                                            value: Str(
                                                "c",
                                            ),
                                            kind: None,
                                        },
                                    },
                                ],
                                values: [
                                    Located {
                                        location: Location {
                                            row: 1,
                                            column: 19,
                                        },
                                        end_location: Some(
                                            Location {
                                                row: 1,
                                                column: 20,
                                            },
                                        ),
                                        custom: (),
                                        node: Constant {
                                            value: Int(
                                                1,
                                            ),
                                            kind: None,
                                        },
                                    },
                                ],
                            },
                        },
                    ],
                },
            },
        },
    },
]
```

File: src/ast/comparable.rs [hunk 3]
```diff
@@ -424,7 +424,11 @@ impl<'a> From<&'a Expr> for ComparableExpr<'a> {
                 orelse: orelse.into(),
             },
             ExprKind::Dict { keys, values } => Self::Dict {
-                keys: keys.iter().map(std::convert::Into::into).collect(),
+                keys: keys
+                    .iter()
+                    .flatten()
+                    .map(std::convert::Into::into)
+                    .collect(),
```
Comment (charliermarsh) [inline src/ast/comparable.rs:431]: I think this may need to preserve the `None`, to be truly comparable, right?

File: src/rules/pyupgrade/rules/unpack_list_comprehension.rs [hunk 4]
```diff
@@ -18,7 +18,11 @@ fn contains_await(expr: &Expr) -> bool {
         ExprKind::IfExp { test, body, orelse } => {
             contains_await(test) || contains_await(body) || contains_await(orelse)
         }
-        ExprKind::Dict { keys, values } => keys.iter().chain(values.iter()).any(contains_await),
+        ExprKind::Dict { keys, values } => keys
+            .iter()
+            .flatten()
+            .chain(values.iter())
+            .any(contains_await),
```
Comment (charliermarsh) [inline src/rules/pyupgrade/rules/unpack_list_comprehension.rs:25]: Oh, we should rewrite this to use `any_over_expr`. Can be a separate PR, or included here...

File: src/ast/comparable.rs (hunk 3 above)
Comment (charliermarsh) [reply src/ast/comparable.rs:431]: (So we'd need to update the definition of `ComparableExpr`.)

Comment (charliermarsh) [reply src/ast/comparable.rs:431]: (I'll go ahead and do this, then merge.)

Comment (harupy) [reply src/ast/comparable.rs:430]: thanks for the update!
